{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PW4: Genetic algorithms for optimization\n",
    "*Alexandra Korukova, Samuel Mayor*  \n",
    "*HEIG-VD - MLG - 30.05.2019*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import sin, cos, tan, arctan, arctan2, arccos, pi\n",
    "import numpy as np\n",
    "import math\n",
    "from pyevolve import G1DList\n",
    "from pyevolve import GSimpleGA\n",
    "from pyevolve import GAllele\n",
    "from pyevolve import Mutators\n",
    "from pyevolve import Initializators\n",
    "from pyevolve import DBAdapters\n",
    "from pyevolve import Crossovers\n",
    "from pyevolve import Consts\n",
    "import sys, random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "LAT = [16.47, 16.47, 20.09, 22.39, 25.23, 22.00, 20.47, 17.20, 16.30, 14.05, 16.53, 21.52, 19.41, 20.09]\n",
    "\n",
    "LON = [96.10, 94.44, 92.54, 93.37, 97.24, 96.05, 97.02, 96.29, 97.38, 98.12, 97.38, 95.59, 97.13, 94.55]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tour_length(matrix, tour):\n",
    "   \"\"\" Returns the total length of the tour \"\"\"\n",
    "   total=0\n",
    "   num_cities=len(tour)\n",
    "   for i in range(num_cities):\n",
    "      j=(i+1)%num_cities\n",
    "      city_i=tour[i]\n",
    "      city_j=tour[j]\n",
    "      total+=matrix[city_i,city_j]\n",
    "   return total\n",
    "\n",
    "def G1DListTSPInitializator(genome, **args):\n",
    "   \"\"\" The initializator for the TSP \"\"\"\n",
    "   genome.clearList()\n",
    "   lst = [i for i in xrange(genome.getListSize())]\n",
    "\n",
    "   for i in xrange(genome.getListSize()):\n",
    "      choice = random.choice(lst)\n",
    "      lst.remove(choice)\n",
    "      genome.append(choice)\n",
    "\n",
    "spherical_distances = []\n",
    "coords = []\n",
    "\n",
    "def eval_func(chromosome):\n",
    "   \"\"\" The evaluation function \"\"\"\n",
    "   global spherical_distances\n",
    "   return tour_length(spherical_distances, chromosome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# source : https://www.johndcook.com/blog/python_longitude_latitude/\n",
    "def spherical_distance(lat1, long1, lat2, long2):\n",
    "\n",
    "    # Convert latitude and longitude to\n",
    "    # spherical coordinates in radians.\n",
    "    degrees_to_radians = math.pi/180.0\n",
    "\n",
    "    # phi = 90 - latitude\n",
    "    phi1 = (90.0 - lat1)*degrees_to_radians\n",
    "    phi2 = (90.0 - lat2)*degrees_to_radians\n",
    "\n",
    "    # theta = longitude\n",
    "    theta1 = long1*degrees_to_radians\n",
    "    theta2 = long2*degrees_to_radians\n",
    "\n",
    "    # Compute spherical distance from spherical coordinates.\n",
    "\n",
    "    # For two locations in spherical coordinates\n",
    "    # (1, theta, phi) and (1, theta', phi')\n",
    "    # cosine( arc length ) =\n",
    "    # sin phi sin phi' cos(theta-theta') + cos phi cos phi'\n",
    "    # distance = rho * arc length\n",
    "\n",
    "    cos = (math.sin(phi1)*math.sin(phi2)*math.cos(theta1 - theta2) +\n",
    "    math.cos(phi1)*math.cos(phi2))\n",
    "    arc = math.acos( cos )\n",
    "\n",
    "    # Remember to multiply arc by the radius of the earth\n",
    "    # in your favorite set of units to get length.\n",
    "    return arc*6371"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gen. 0 (0.00%): Max/Min/Avg Fitness(Raw) [8118.14(8290.22)/5318.98(5135.05)/6765.11(6765.11)]\n",
      "Gen. 100 (10.00%): Max/Min/Avg Fitness(Raw) [5416.20(6125.86)/4370.82(4258.66)/4513.50(4513.50)]\n",
      "Gen. 200 (20.00%): Max/Min/Avg Fitness(Raw) [5280.12(5919.38)/4243.96(4130.53)/4400.10(4400.10)]\n",
      "Gen. 300 (30.00%): Max/Min/Avg Fitness(Raw) [4759.95(6028.10)/3818.13(3580.75)/3966.63(3966.63)]\n",
      "Gen. 400 (40.00%): Max/Min/Avg Fitness(Raw) [4490.50(5880.57)/3619.63(3392.19)/3742.08(3742.08)]\n",
      "Gen. 500 (50.00%): Max/Min/Avg Fitness(Raw) [4307.83(6085.22)/3532.99(3392.19)/3589.86(3589.86)]\n",
      "Gen. 600 (60.00%): Max/Min/Avg Fitness(Raw) [4327.23(5965.76)/3540.67(3392.19)/3606.02(3606.02)]\n",
      "Gen. 700 (70.00%): Max/Min/Avg Fitness(Raw) [4498.68(6396.77)/3647.89(3392.19)/3748.90(3748.90)]\n",
      "Gen. 800 (80.00%): Max/Min/Avg Fitness(Raw) [4430.42(5584.10)/3575.00(3392.19)/3692.01(3692.01)]\n",
      "Gen. 900 (90.00%): Max/Min/Avg Fitness(Raw) [4713.78(7040.13)/3792.84(3392.19)/3928.15(3928.15)]\n",
      "Gen. 1000 (100.00%): Max/Min/Avg Fitness(Raw) [4549.53(6696.93)/3687.13(3392.19)/3791.28(3791.28)]\n",
      "Total time elapsed: 9.466 seconds.\n",
      "- GenomeBase\n",
      "\tScore:\t\t\t 3392.187220\n",
      "\tFitness:\t\t 3687.129956\n",
      "\n",
      "\tParams:\t\t {'allele': - GAlleles\n",
      "\tHomogeneous:\t True\n",
      "\tList size:\t 1\n",
      "\tAlleles:\n",
      "\n",
      "Allele for 0 position:\n",
      "- GAlleleList\n",
      "\tList size:\t 14\n",
      "\tAllele Options:\t [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]\n",
      "\n",
      "}\n",
      "\n",
      "\tSlot [Evaluator] (Count: 1)\n",
      "\t\tName: eval_func - Weight: 0.50\n",
      "\t\tDoc:  The evaluation function \n",
      "\tSlot [Initializator] (Count: 1)\n",
      "\t\tName: G1DListTSPInitializator - Weight: 0.50\n",
      "\t\tDoc:  The initializator for the TSP \n",
      "\tSlot [Mutator] (Count: 1)\n",
      "\t\tName: G1DListMutatorSwap - Weight: 0.50\n",
      "\t\tDoc:  The mutator of G1DList, Swap Mutator\n",
      "   \n",
      "   .. note:: this mutator is :term:`Data Type Independent`\n",
      "\n",
      "   \n",
      "\tSlot [Crossover] (Count: 1)\n",
      "\t\tName: G1DListCrossoverOX - Weight: 0.50\n",
      "\t\tDoc:  The OX Crossover for G1DList  (order crossover) \n",
      "\n",
      "- G1DList\n",
      "\tList size:\t 14\n",
      "\tList:\t\t [0, 7, 12, 6, 11, 5, 4, 3, 2, 13, 1, 9, 8, 10]\n",
      "\n",
      "\n",
      "3392.1872196164536\n"
     ]
    }
   ],
   "source": [
    "# creates latitude/longitude tuples for each city\n",
    "coords = []\n",
    "for i in range(len(LAT)):\n",
    "    coords.append((LAT[i], LON[i]))\n",
    "\n",
    "# calculate distances between each city pairs\n",
    "spherical_distances = np.zeros((14, 14))\n",
    "for i,c1 in enumerate(coords):\n",
    "    for j,c2 in enumerate(coords):\n",
    "        spherical_distances[i, j] = spherical_distance(c1[0], c1[1], c2[0], c2[1])\n",
    "\n",
    "# set the alleles to the cities numbers\n",
    "setOfAlleles = GAllele.GAlleles(homogeneous=True)\n",
    "lst = [ i for i in xrange(len(coords)) ]\n",
    "a = GAllele.GAlleleList(lst)\n",
    "setOfAlleles.add(a)\n",
    "\n",
    "genome = G1DList.G1DList(len(coords))\n",
    "genome.setParams(allele=setOfAlleles)\n",
    "\n",
    "genome.evaluator.set(eval_func)\n",
    "genome.mutator.set(Mutators.G1DListMutatorSwap)\n",
    "genome.crossover.set(Crossovers.G1DListCrossoverOX)\n",
    "genome.initializator.set(G1DListTSPInitializator)\n",
    "\n",
    "for i in np.arrange(100, 1001, 100):\n",
    "    ga = GSimpleGA.GSimpleGA(genome)\n",
    "    ga.setGenerations(i)\n",
    "    ga.setMinimax(Consts.minimaxType[\"minimize\"])\n",
    "    ga.setCrossoverRate(1.0)\n",
    "    ga.setMutationRate(0.03)\n",
    "    ga.setPopulationSize(80)\n",
    "\n",
    "    ga.evolve(freq_stats=100)\n",
    "    best = ga.bestIndividual()\n",
    "    print best\n",
    "    print(tour_length(spherical_distances, best))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0.00000000e+00 1.77009305e+02 5.50675649e+02 7.17757240e+02\n",
      "  9.81224413e+02 6.14930327e+02 4.55234726e+02 8.36531369e+01\n",
      "  1.37851129e+02 3.45482052e+02 1.36631089e+02 5.64086509e+02\n",
      "  3.44588856e+02 4.34508524e+02]\n",
      " [1.77009305e+02 0.00000000e+00 4.49724987e+02 6.67756396e+02\n",
      "  1.01647667e+03 6.37693019e+02 5.21372205e+02 2.12968085e+02\n",
      "  3.14202703e+02 4.77725171e+02 3.13518919e+02 5.74393117e+02\n",
      "  4.33391525e+02 4.02693079e+02]\n",
      " [5.50675649e+02 4.49724987e+02 0.00000000e+00 2.69825166e+02\n",
      "  7.47657237e+02 4.21626921e+02 4.69163853e+02 5.09234088e+02\n",
      "  6.62466452e+02 8.95787809e+02 6.46241876e+02 3.54657083e+02\n",
      "  4.86258069e+02 2.09901390e+02]\n",
      " [7.17757240e+02 6.67756396e+02 2.69825166e+02 0.00000000e+00\n",
      "  5.04643144e+02 2.79304235e+02 4.33921938e+02 6.52907115e+02\n",
      "  7.97079289e+02 1.05408788e+03 7.75318270e+02 2.48544627e+02\n",
      "  5.12152028e+02 2.83479792e+02]\n",
      " [9.81224413e+02 1.01647667e+03 7.47657237e+02 5.04643144e+02\n",
      "  9.49352980e-05 3.79063065e+02 5.29767283e+02 8.98296394e+02\n",
      "  9.93077069e+02 1.24655563e+03 9.67504897e+02 4.45568206e+02\n",
      "  6.47253255e+02 6.34641627e+02]\n",
      " [6.14930327e+02 6.37693019e+02 4.21626921e+02 2.79304235e+02\n",
      "  3.79063065e+02 0.00000000e+00 1.97611103e+02 5.34326937e+02\n",
      "  6.49008546e+02 9.10637517e+02 6.24036341e+02 7.14523639e+01\n",
      "  3.09122804e+02 2.63315547e+02]\n",
      " [4.55234726e+02 5.21372205e+02 4.69163853e+02 4.33921938e+02\n",
      "  5.29767283e+02 1.97611103e+02 0.00000000e+00 3.71632032e+02\n",
      "  4.65235362e+02 7.23351708e+02 4.39748735e+02 1.88861695e+02\n",
      "  1.18426109e+02 2.61064918e+02]\n",
      " [8.36531369e+01 2.12968085e+02 5.09234088e+02 6.52907115e+02\n",
      "  8.98296394e+02 5.34326937e+02 3.71632032e+02 0.00000000e+00\n",
      "  1.53246969e+02 4.01342403e+02 1.37853958e+02 4.85939236e+02\n",
      "  2.61248722e+02 3.69954282e+02]\n",
      " [1.37851129e+02 3.14202703e+02 6.62466452e+02 7.97079289e+02\n",
      "  9.93077069e+02 6.49008546e+02 4.65235362e+02 1.53246969e+02\n",
      "  0.00000000e+00 2.62488256e+02 2.55748331e+01 6.10188357e+02\n",
      "  3.46826694e+02 5.16649585e+02]\n",
      " [3.45482052e+02 4.77725171e+02 8.95787809e+02 1.05408788e+03\n",
      "  1.24655563e+03 9.10637517e+02 7.23351708e+02 4.01342403e+02\n",
      "  2.62488256e+02 0.00000000e+00 2.86956642e+02 8.72676603e+02\n",
      "  6.05248287e+02 7.71296618e+02]\n",
      " [1.36631089e+02 3.13518919e+02 6.46241876e+02 7.75318270e+02\n",
      "  9.67504897e+02 6.24036341e+02 4.39748735e+02 1.37853958e+02\n",
      "  2.55748331e+01 2.86956642e+02 0.00000000e+00 5.85874243e+02\n",
      "  3.21330931e+02 4.95894819e+02]\n",
      " [5.64086509e+02 5.74393117e+02 3.54657083e+02 2.48544627e+02\n",
      "  4.45568206e+02 7.14523639e+01 1.88861695e+02 4.85939236e+02\n",
      "  6.10188357e+02 8.72676603e+02 5.85874243e+02 0.00000000e+00\n",
      "  2.84220747e+02 1.92273133e+02]\n",
      " [3.44588856e+02 4.33391525e+02 4.86258069e+02 5.12152028e+02\n",
      "  6.47253255e+02 3.09122804e+02 1.18426109e+02 2.61248722e+02\n",
      "  3.46826694e+02 6.05248287e+02 3.21330931e+02 2.84220747e+02\n",
      "  0.00000000e+00 2.80390111e+02]\n",
      " [4.34508524e+02 4.02693079e+02 2.09901390e+02 2.83479792e+02\n",
      "  6.34641627e+02 2.63315547e+02 2.61064918e+02 3.69954282e+02\n",
      "  5.16649585e+02 7.71296618e+02 4.95894819e+02 1.92273133e+02\n",
      "  2.80390111e+02 0.00000000e+00]]\n"
     ]
    }
   ],
   "source": [
    "print(spherical_distances)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "- GSimpleGA\n",
      "\tGP Mode:\t\t False\n",
      "\tPopulation Size:\t 80\n",
      "\tGenerations:\t\t 1000\n",
      "\tCurrent Generation:\t 1000\n",
      "\tMutation Rate:\t\t 0.03\n",
      "\tCrossover Rate:\t\t 1.00\n",
      "\tMinimax Type:\t\t Minimize\n",
      "\tElitism:\t\t True\n",
      "\tElitism Replacement:\t 1\n",
      "\tDB Adapter:\t\t None\n",
      "\tSlot [Selector] (Count: 1)\n",
      "\t\tName: GRankSelector - Weight: 0.50\n",
      "\t\tDoc:  The Rank Selector - This selector will pick the best individual of\n",
      "   the population every time.\n",
      "   \n",
      "\tSlot [Generation Step Callback] (Count: 0)\n",
      "\t\tNo function\n",
      "\tSlot [Termination Criteria] (Count: 0)\n",
      "\t\tNo function\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(ga)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
